File: Number Theory/leetcode_1201.py

# Problem link ===>> https://leetcode.com/problems/ugly-number-iii/description/?envType=problem-list-v2&envId=21r1b7y6
# submission link == =>> https://leetcode.com/problems/ugly-number-iii/submissions/1758345421/?envType=problem-list-v2&envId=21r1b7y6

import sys
import math
import bisect
import heapq
from collections import defaultdict, deque, OrderedDict, Counter
from functools import lru_cache, reduce
from itertools import permutations, combinations, accumulate
from typing import List, Tuple, Dict, Set, Optional, Union, Any

MOD = 10**9 + 7
INF = float('inf')
input = sys.stdin.read
print = sys.stdout.write


class Solution:
    def nthUglyNumber(self, n: int, a: int, b: int, c: int) -> int:
        def gcd(x, y):
            while y:
                x, y = y, x % y
            return x

        def lcm(x, y):
            return x * y // gcd(x, y)

        def count(x):
            ab = lcm(a, b)
            ac = lcm(a, c)
            bc = lcm(b, c)
            abc = lcm(ab, c)
            return x // a + x // b + x // c - x // ab - x // ac - x // bc + x // abc

        left, right = 1, 2 * 10**9
        while left < right:
            mid = (left + right) // 2
            if count(mid) < n:
                left = mid + 1
            else:
                right = mid
        return left

def main():
    t = int(input())
    for _ in range(t):
        n = int(input())

if __name__ == "__main__":
    solution = Solution()

File: Number Theory/leetcode_1201.cpp

// Problem link ===>> https://leetcode.com/problems/ugly-number-iii/description/?envType=problem-list-v2&envId=21r1b7y6

#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace std;
using namespace __gnu_pbds;
#define ll long long
#define ld long double
#define endl '\n'
const int MOD = 1e9 + 7;
const double EPS = 1e-9;
const double PI = acos(-1.0);
template <typename T>
using ordered_set = tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;
template <typename K, typename V>
using ordered_map = tree<
    K, V, less<K>, rb_tree_tag, tree_order_statistics_node_update>;

class Solution
{
public:
    int nthUglyNumber(int n, int a, int b, int c)
    {
        auto gcd = [](ll x, ll y) {
            while (y) {
                x %= y;
                swap(x, y);
            }
            return x;
        };

        auto lcm = [&](ll x, ll y) {
            return (x / gcd(x, y)) * y;
        };

        ll ab = lcm(a, b);
        ll ac = lcm(a, c);
        ll bc = lcm(b, c);
        ll abc = lcm(ab, c);

        auto count = [&](ll x) {
            return (x / a) + (x / b) + (x / c) - (x / ab) - (x / ac) - (x / bc) + (x / abc);
        };

        ll left = 1, right = 2e9;
        while (left < right) {
            ll mid = left + (right - left) / 2;
            if (count(mid) >= n)
                right = mid;
            else
                left = mid + 1;
        }

        return left;
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);
    
    ll t;
    cin >> t;
    while (t--)
    {
        ll n;
        cin >> n;

        
    }
    return 0;
}